pennylane/qnn/torch.py

```diff
@@ -222,41 +253,51 @@ def __init__(self, qnode, weight_shapes: dict, init_method: Optional[Callable] =
                 "weight_shapes".format(self.input_arg)
             )
 
-        if set(weight_shapes.keys()) | {self.input_arg} != set(self.sig.keys()):
-            raise ValueError("Must specify a shape for every non-input parameter in the QNode")
+        param_kinds = [p.kind for p in self.sig.values()]
+
+        if inspect.Parameter.VAR_POSITIONAL in param_kinds:
+            raise TypeError("Cannot have a variable number of positional arguments")
+
+        if inspect.Parameter.VAR_KEYWORD not in param_kinds:
+            if set(weight_shapes.keys()) | {self.input_arg} != set(self.sig.keys()):
+                raise ValueError("Must specify a shape for every non-input parameter in the QNode")
```

Comment:
I was imagining another use case where a user wants to pass options via a variable number of keyword arguments, e.g.
`qnode(inputs, weights, **options)`, but that wouldn't be supported here unless we updated `forward()`.

Reply:
> If there isn't a variable number of keyword arguments, we make sure that all shapes are specified?
> Otherwise, we allow for the possibility of a variable number of keyword arguments?

Yeah, essentially! Feel free to push directly and change this logic. Essentially, it's just acknowledging that `set(self.sig.keys())` is not well defined if `**kwargs` is present; we can never inspect the number of expected non-input arguments, we just have to assume the user knows what they are doing.

> `qnode(inputs, weights, **options)`, but that wouldn't be supported here unless we updated `forward()`.

This would be great to support, do you think it's doable? Feel free to give it a shot if you want!

I also tried to get `*args` to work, but realised we are restricted by `weight_shapes`, since it is a mapping from argument name to shape. I figure `*args` would only work if `weight_shapes` was a list of shapes?

Reply:
Sounds good! My only thought would be how/whether we need to update the docs, since we mention not having a variable number of kwargs as a restriction. Maybe as a TODO in the code that we'll see when removing the old QNode?

Reply:
Agreed! I didn't update the docs here, as I thought it would be two confusing to have _two_ different sets of restrictions depending on the mode. I'll add a TODO in the first tape mode branch.